### PR TITLE
Make ProbeTimeout and ProbeBackoff Configurable

### PR DESCRIPTION
**Pull request by xqi-nr**
# Description
there is a parameter we use to control the timeout of probing infra agent, which is hardcoded as 90s, during scraper pod starts. This timeout allow 3 attempts to probe the infra agent during the container starting. If none of these 3 attempts succeeds, the scraper pod will restart. A customer is complaining the pod restart due to probe timeout.

# PR Changes
- Make probe timeout during scraper pod starts configurable in order to give more controls to customer regarding this restarting issue.
- Udpate readme

# Tests 
cc @htroisi

In my values-local.yaml, I make the following changes
```
cluster: minikube
global:
  licenseKey: <my staging license key>
  nrStaging: true
sink:
  http:
    # probeTimeout is the amount of time the scraper container to probe infra agent sidecar container before giving up and restarting during pod starts.
    probeTimeout: 90s
```
 
The e2e tests succeeded means all the agents works as expected and can send data to NR backend.

Locally, I run `make local-env-start` and check the deployed agent status as follows

```
xqi@TFXCKKXMP9 nri-kubernetes % k get po | grep nr-nrk8s
nr-nrk8s-controlplane-qvlfg                        2/2     Running                0          2m34s
nr-nrk8s-ksm-ddb45fd78-hbrrt                       2/2     Running                0          2m34s
nr-nrk8s-kubelet-jmgsx                             2/2     Running                0          2m34s
xqi@TFXCKKXMP9 nri-kubernetes % k logs nr-nrk8s-kubelet-jmgsx       
Defaulted container "kubelet" out of: kubelet, agent
time="2023-03-14T19:23:20Z" level=info msg="Waiting for agent container to be ready..."
time="2023-03-14T19:23:40Z" level=info msg="New Relic Kubernetes integration Version: dev, Platform: linux/amd64, GoVersion: go1.20, GitCommit: bdc8cb52f3b6694d91e8c52c33d68015bf09bc2e, BuildDate: Tue Mar 14 12:23:02 PDT 2023\n"
time="2023-03-14T19:23:40Z" level=info msg="Trying to connect to kubelet locally with scheme=\"https\" hostURL=\"192.168.67.2:10250\""
time="2023-03-14T19:23:40Z" level=info msg="Connected to Kubelet through nodeIP with scheme=\"https\" hostURL=\"192.168.67.2:10250\""
xqi@TFXCKKXMP9 nri-kubernetes % k describe po nr-nrk8s-kubelet-jmgsx
Name:             nr-nrk8s-kubelet-jmgsx
Namespace:        default
Priority:         0
Service Account:  nr-newrelic-infrastructure
Node:             minikube/192.168.67.2
Start Time:       Tue, 14 Mar 2023 12:23:17 -0700
Labels:           app.kubernetes.io/component=kubelet
...
Annotations:      checksum/agent-config: 3c57d1534528615a326f66ac634a015aeac2bffc2f7382cc2560dee9af906e1f
...
Status:           Running
IP:               10.244.0.114
IPs:
  IP:           10.244.0.114
Controlled By:  DaemonSet/nr-nrk8s-kubelet
Containers:
  kubelet:
    Container ID:  docker://9673f431248aa099f17ee639023901a1a782d81e4bc0bf157b5266936dd4474c
    Image:         nri-kubernetes:tilt-35034f08c1545fee
    Image ID:      docker://sha256:35034f08c1545fee66c20f2092dcc7c2564f13c09b81d2462f2ffb46d36d752f
    Port:          <none>
    Host Port:     <none>
    Command:
      /tilt-restart-wrapper
      --watch_file=/tmp/.restart-proc
      /usr/local/bin/nri-kubernetes
    State:          Running
      Started:      Tue, 14 Mar 2023 12:23:19 -0700
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  300M
    Requests:
      cpu:     100m
      memory:  150M
    Environment:
      NRI_KUBERNETES_SINK_HTTP_PORT:  8003
      NRI_KUBERNETES_CLUSTERNAME:     minikube
      NRI_KUBERNETES_VERBOSE:         false
      NRI_KUBERNETES_NODENAME:         (v1:spec.nodeName)
      NRI_KUBERNETES_NODEIP:           (v1:status.hostIP)
    Mounts:
      /etc/newrelic-infra/nri-kubernetes.yml from nri-kubernetes-config (rw,path="nri-kubernetes.yml")
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-b2h8j (ro)
  agent:
    Container ID:  docker://9d292a7d9594580d603721b13fd395f751c4128b3f1792cd6599e680bd2d5af5
    Image:         newrelic/infrastructure-bundle:2.8.38
    Image ID:      docker-pullable://newrelic/infrastructure-bundle@sha256:f9b5c9190cd130f15f1266fc008e06dc503d9b209f012625eac40c04ad0c28a9
    Port:          8003/TCP
    Host Port:     0/TCP
    Args:
      newrelic-infra
    State:          Running
      Started:      Tue, 14 Mar 2023 12:23:20 -0700
    Ready:          True
    Restart Count:  0
    Limits:
      memory:  300M
    Requests:
      cpu:     100m
      memory:  150M
    Environment:
      NRIA_LICENSE_KEY:              <set to the key 'licenseKey' in secret 'nr-newrelic-infrastructure-license'>  Optional: false
      NRIA_OVERRIDE_HOSTNAME_SHORT:   (v1:spec.nodeName)
      NRIA_OVERRIDE_HOSTNAME:         (v1:spec.nodeName)
      NRI_KUBERNETES_NODE_NAME:       (v1:spec.nodeName)
      CLUSTER_NAME:                  minikube
      NRIA_PASSTHROUGH_ENVIRONMENT:  CLUSTER_NAME
      NRIA_HOST:                      (v1:status.hostIP)
    Mounts:
      /dev from dev (rw)
      ...
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  dev:
    Type:          HostPath (bare host directory volume)
    Path:          /dev
    HostPathType:  
 ...
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 :NoSchedule op=Exists
                             :NoExecute op=Exists
                             node.kubernetes.io/disk-pressure:NoSchedule op=Exists
...
Events:
  Type    Reason     Age    From               Message
  ----    ------     ----   ----               -------
  Normal  Scheduled  2m52s  default-scheduler  Successfully assigned default/nr-nrk8s-kubelet-jmgsx to minikube
  Normal  Pulled     2m51s  kubelet            Container image "nri-kubernetes:tilt-35034f08c1545fee" already present on machine
  Normal  Created    2m51s  kubelet            Created container kubelet
  Normal  Started    2m51s  kubelet            Started container kubelet
  Normal  Pulled     2m51s  kubelet            Container image "newrelic/infrastructure-bundle:2.8.38" already present on machine
  Normal  Created    2m51s  kubelet            Created container agent
  Normal  Started    2m50s  kubelet            Started container agent
```

### CODE REVIEW

File: charts/newrelic-infrastructure/values.yaml [hunk 1]
```diff
@@ -113,13 +113,17 @@ kubelet:
     requests:
       cpu: 100m
       memory: 150M
+  # -- Max number of scraper rerun after scraper runtime error happens
+  scraperMaxReruns: 4
+  sink:
+    http:
+      # probeTimeout is the amount of time the main func to probe infra agent sidecar before giving up and restarting during scraper agent starting.
+      probeTimeout: 90s
```
Comment (htroisi) [inline charts/newrelic-infrastructure/values.yaml:121]: Should these be added to the chart README?

Comment (xqi-nr) [reply charts/newrelic-infrastructure/values.yaml:121]: Good point, I will add them to the readme

File: charts/newrelic-infrastructure/README.md [hunk 2]
```diff
@@ -114,6 +114,7 @@ integrations that you have configured.
 | common.config.interval | duration | `15s` (See [Low data mode](README.md#low-data-mode)) | Intervals larger than 40s are not supported and will cause the NR UI to not behave properly. Any non-nil value will override the `lowDataMode` default. |
 | common.config.namespaceSelector | object | `{}` | Config for filtering ksm and kubelet metrics by namespace. |
 | containerSecurityContext | object | `{}` | Sets security context (at container level). Can be configured also with `global.containerSecurityContext` |
+| sink.http.probeTimeout | time.duration | 90s | The amount of time the scraper container to probe infra agent sidecar container before giving up and restarting during scraper agent starting. |
```
Comment (htroisi) [inline charts/newrelic-infrastructure/README.md:117]: It looks like the rest of these configs are alphabetical. Should this one be in alphabetical order? I think there is a `helm-docs` tool that people use to automatically do this for you.

Comment (aimichelle) [inline charts/newrelic-infrastructure/README.md:117]: Nit: from looking at the other rows, it looks like the styling for "90s" should be \`90s\` (see `common.config.interval` above).

File: src/integration/wrapper.go [hunk 3]
```diff
@@ -69,7 +69,12 @@ func WithHTTPSink(sinkConfig config.HTTPSink) OptionFunc {
 			}
 		}
 
-		prober, err := prober.New(iw.probeTimeout, iw.probeBackoff, prober.WithLogger(iw.logger), prober.WithClient(client))
+		probeTimeout := sinkConfig.ProbeTimeout
+		// if the ProbeTimeout is not set in HTTPSink, just use the default probe timeout from the wrapper
+		if probeTimeout == 0*time.Second {
```
Comment (aimichelle) [inline src/integration/wrapper.go:74]: You can use `time.IsZero()` here.

File: charts/newrelic-infrastructure/README.md (hunk 2 above)
Comment (xqi-nr) [reply charts/newrelic-infrastructure/README.md:117]: thanks. move it to be alphabetical order

Comment (xqi-nr) [reply charts/newrelic-infrastructure/README.md:117]: thanks. updated

File: src/integration/wrapper.go (hunk 3 above)
Comment (xqi-nr) [reply src/integration/wrapper.go:74]: probeTimeout is of time.Duration. I cannot use time.IsZero() here.
I want to check whether it is set or not here. If not set, then probeTimeout should be zero duration

File: src/kubelet/scraper.go [hunk 4]
```diff
@@ -148,5 +148,5 @@ func (s *Scraper) IncCurrentReruns() {
 
 // Check whether the max number of kubulet scraper reruns has been reached or not.
 func (s *Scraper) IsMaxRerunReached() bool {
-	return s.currentReruns > s.config.ScraperMaxReruns
+	return s.currentReruns > s.config.Kubelet.ScraperMaxReruns
```
Comment (vihangm) [inline src/kubelet/scraper.go:151]: can you pull this fix out into a separate PR?

Comment (xqi-nr) [reply src/kubelet/scraper.go:151]: Sure I will make a separate PR for this.

BTW, This is not a fix. It just changes the way the code refers to ScraperMaxReruns, which seems more explicit.

Comment (xqi-nr) [reply src/kubelet/scraper.go]: pulled out

File: src/integration/wrapper.go (hunk 3 above)
Comment (vihangm) [reply src/integration/wrapper.go:74]: you can probably use `0` on the RHS. instead of using `*time.Second`
The compiler should use the LHS type to automatically use a time.Duration typed zero value.

**Review by vihangm — APPROVED** [review]

File: src/integration/wrapper.go (hunk 3 above)
Comment (xqi-nr) [reply src/integration/wrapper.go]: done

File: src/integration/wrapper.go [hunk 5]
```diff
@@ -69,7 +69,12 @@ func WithHTTPSink(sinkConfig config.HTTPSink) OptionFunc {
 			}
 		}
 
-		prober, err := prober.New(iw.probeTimeout, iw.probeBackoff, prober.WithLogger(iw.logger), prober.WithClient(client))
+		probeTimeout := sinkConfig.ProbeTimeout
+		// if the ProbeTimeout is not set in HTTPSink, just use the default probe timeout from the wrapper
+		if probeTimeout == 0 {
+			probeTimeout = iw.probeTimeout
+		}
```
Comment (vihangm) [inline src/integration/wrapper.go:76]: can you remove the `probeTimeout` on the Wrapper struct in this file?
also remove `defaultProbeTimeout` from this file.

it seems like those options only apply to the HTTPSink, and cannot be overridden anyway. so it's strange that they are defined in the wrapper struct.

as a followup, we should also move the probe backoff into an HTTPSink option and remove it from the wrapper.

**Review by vihangm — APPROVED** [review]
